### PR TITLE
[websocket-nio]Close connection for invalid UTF-8 payload data

### PR DESCRIPTION
**Pull request by nethraravindran**
The connection must be dropped/closed with error code 1007 if the payload is not a valid UTF-8 encoded data

### CODE REVIEW

File: Sources/KituraWebSocket/WebSocketConnection.swift [hunk 1]
```diff
@@ -105,9 +105,13 @@ extension WebSocketConnection: ChannelInboundHandler {
...
                 if frame.fin {
-                    var data = unmaskedData(frame: frame)
-                    let text = data.readString(length: data.readableBytes) ?? ""
-                    fireReceivedString(message: text)
+                    let data = unmaskedData(frame: frame)
+                    if let text = data.getString(at: 0, length: data.readableBytes, encoding: .utf8) {
+                        fireReceivedString(message: text)
+                    } else {
+                        connectionClosed(reason: .dataInconsistentWithMessage, description: "Failed to convert received payload to UTF-8 String")
+                        return
```
Comment (pushkarnk) [inline Sources/KituraWebSocket/WebSocketConnection.swift]: Is this `return` needed?

File: Sources/KituraWebSocket/WebSocketConnection.swift [hunk 2]
```diff
@@ -144,7 +148,12 @@ extension WebSocketConnection: ChannelInboundHandler {
                     case .binary:
                         fireReceivedData(data: frame.data.getData(at: 0, length: frame.data.readableBytes) ?? Data())
                     case .text:
-                        fireReceivedString(message: frame.data.getString(at: 0, length: frame.data.readableBytes) ?? "") 
+                        if let text = frame.unmaskedData.getString(at: 0, length: frame.unmaskedData.readableBytes, encoding: .utf8) {
+                            fireReceivedString(message: text)
+                        } else {
+                            connectionClosed(reason: .dataInconsistentWithMessage, description: "Failed to convert received payload to UTF-8 String")
+                            return
```
Comment (pushkarnk) [inline Sources/KituraWebSocket/WebSocketConnection.swift]: Is this `return` needed?

File: Sources/KituraWebSocket/WebSocketConnection.swift [hunk 3]
```diff
@@ -105,9 +105,13 @@ extension WebSocketConnection: ChannelInboundHandler {
...
                 
                 if frame.fin {
-                    var data = unmaskedData(frame: frame)
-                    let text = data.readString(length: data.readableBytes) ?? ""
-                    fireReceivedString(message: text)
+                    let data = unmaskedData(frame: frame)
+                    if let text = data.getString(at: 0, length: data.readableBytes, encoding: .utf8) {
+                        fireReceivedString(message: text)
+                    } else {
+                        connectionClosed(reason: .dataInconsistentWithMessage, description: "Failed to convert received payload to UTF-8 String")
```
Comment (pushkarnk) [inline Sources/KituraWebSocket/WebSocketConnection.swift]: The Kitura-WebSocket tests expect this error code to be 1005 i.e `WebSocketErrorCode(codeNumber: 1005)`

File: Sources/KituraWebSocket/WebSocketConnection.swift [hunk 4]
```diff
@@ -144,7 +148,12 @@ extension WebSocketConnection: ChannelInboundHandler {
                     case .binary:
                         fireReceivedData(data: frame.data.getData(at: 0, length: frame.data.readableBytes) ?? Data())
                     case .text:
-                        fireReceivedString(message: frame.data.getString(at: 0, length: frame.data.readableBytes) ?? "") 
+                        if let text = frame.unmaskedData.getString(at: 0, length: frame.unmaskedData.readableBytes, encoding: .utf8) {
+                            fireReceivedString(message: text)
+                        } else {
+                            connectionClosed(reason: .dataInconsistentWithMessage, description: "Failed to convert received payload to UTF-8 String")
```
Comment (pushkarnk) [inline Sources/KituraWebSocket/WebSocketConnection.swift]: The Kitura-WebSocket tests expect this error code to be 1005 i.e `WebSocketErrorCode(codeNumber: 1005)`

**Comment by pushkarnk** [comment]
@nethraravindran Can you resolve the merge conflicts please?

**Comment by nethraravindran** [comment]
@pushkarnk  Resolved the conflicts :)

**Pull request merged**